hedera-node/src/main/java/com/hedera/services/ledger/interceptors/StakeAwareAccountsCommitsInterceptor.java

```diff
@@ -0,0 +1,325 @@
+			final RewardCalculator rewardCalculator,
+			final StakeChangeManager stakeChangeManager,
+			final StakePeriodManager stakePeriodManager,
+			final StakeInfoManager stakeInfoManager) {
+		super(sideEffectsTracker);
+		this.stakeChangeManager = stakeChangeManager;
+		this.networkCtx = networkCtx;
+		this.rewardCalculator = rewardCalculator;
+		this.sideEffectsTracker = sideEffectsTracker;
+		this.dynamicProperties = dynamicProperties;
+		this.stakePeriodManager = stakePeriodManager;
+		this.stakeInfoManager = stakeInfoManager;
+	}
+
+	@Override
+	public void preview(final EntityChangeSet<AccountID, MerkleAccount, AccountProperty> pendingChanges) {
+		if (pendingChanges.size() == 0) {
+			return;
+		}
+		// if the rewards are activated previously they will not be activated again
+		rewardsActivated = rewardsActivated || networkCtx.get().areRewardsActivated();
+
+		// Iterate through the change set, maintaining two invariants:
+		//   1. At the beginning of iteration i, any account that is rewardable due to change in balance or
+		//      stakedAccountId or stakedNodeId or declineRewards fields. Also checks the balance of funding account
+		//      0.0.800 if it has reached the ONE TIME threshold to activate staking.
+		//      NOTE that this activation happens only once.
+		//   2. Any account whose stakedToMe balance is affected by a change in the [0, i) range has
+		//      been, if not already present, added to the pendingChanges; and its changes include its
+		//      new STAKED_TO_ME change
+		updateAccountStakes(pendingChanges);
+
+		// Iterate through the change set again to update node stakes; we do this is in a
+		// separate loop to ensure all STAKED_TO_ME fields have their final values
+		updateNodeStakes(pendingChanges);
+		updateFundingRewardBalances(pendingChanges);
+
+		super.preview(pendingChanges);
+
+		if (!rewardsActivated) {
+			checkStakingRewardsActivation(pendingChanges);
+		}
+	}
+
+	private void updateAccountStakes(final EntityChangeSet<AccountID, MerkleAccount, AccountProperty> pendingChanges) {
+		for (int i = 0, n = pendingChanges.size(); i < n; i++) {
+			hasBeenRewarded[i] = false;
+
+			final var account = pendingChanges.entity(i);
+			final var changes = pendingChanges.changes(i);
+
+			// Update BALANCE and STAKE_PERIOD_START in the pending changes for this account, if reward-eligible
+			if (isRewardable(account, changes)) {
+				payReward(i, account, changes);
+			}
+			// Update any STAKED_TO_ME side effects of this change
+			n = updateStakedToMeSideEffects(account, changes, pendingChanges);
+		}
+	}
+
+	private void updateFundingRewardBalances(
+			final EntityChangeSet<AccountID, MerkleAccount, AccountProperty> pendingChanges) {
+		final var rewardsPaid = rewardCalculator.rewardsPaidInThisTxn();
+		if (rewardsPaid > 0) {
+			final var rewardAccountI = stakeChangeManager.findOrAdd(800L, pendingChanges);
+			updateBalance(-rewardsPaid, rewardAccountI, pendingChanges);
```

Comment:
If we make `updateBalance()` return the final balance; then here we know the new final balance for `0.0.800`.

Then we can return the balance from this method, and pass it into `checkStakingRewardsActivation()`. This eliminates the need for `calculateNewRewardBalance()` entirely.